nexus/blueprint-execution/src/resource_allocation.rs

```diff
@@ -0,0 +1,998 @@
+                    allocator
+                        .ensure_nexus_external_networking_allocated(
+                            z.id,
+                            *external_ip,
+                            nic,
+                        )
+                        .await?;
+                }
+                OmicronZoneType::ExternalDns { dns_address, nic, .. } => {
+                    allocator
+                        .ensure_external_dns_external_networking_allocated(
+                            z.id,
+                            dns_address,
+                            nic,
+                        )
+                        .await?;
+                }
+                OmicronZoneType::BoundaryNtp { snat_cfg, nic, .. } => {
+                    allocator
+                        .ensure_boundary_ntp_external_networking_allocated(
+                            z.id, snat_cfg, nic,
+                        )
+                        .await?;
+                }
+                OmicronZoneType::InternalNtp { .. }
+                | OmicronZoneType::Clickhouse { .. }
+                | OmicronZoneType::ClickhouseKeeper { .. }
+                | OmicronZoneType::CockroachDb { .. }
+                | OmicronZoneType::Crucible { .. }
+                | OmicronZoneType::CruciblePantry { .. }
+                | OmicronZoneType::InternalDns { .. }
+                | OmicronZoneType::Oximeter { .. } => (),
+            }
+        }
+    }
+
+    Ok(())
+}
+
+struct ResourceAllocator<'a> {
+    log: &'a Logger,
+    opctx: &'a OpContext,
+    datastore: &'a DataStore,
+}
+
+impl<'a> ResourceAllocator<'a> {
+    // Helper function to determine whether a given external IP address is
+    // already allocated to a specific service zone.
+    async fn is_external_ip_already_allocated(
+        &self,
+        zone_type: &'static str,
+        zone_id: Uuid,
+        external_ip: IpAddr,
+        port_range: Option<(u16, u16)>,
+    ) -> anyhow::Result<bool> {
+        let allocated_ips = self
+            .datastore
+            .service_lookup_external_ips(self.opctx, zone_id)
+            .await
+            .with_context(|| {
+                format!(
+                    "failed to look up external IPs for {zone_type} {zone_id}"
+                )
+            })?;
+
+        if !allocated_ips.is_empty() {
+            // All the service zones that want external IP addresses only expect
+            // to have a single IP. This service already has (at least) one:
+            // make sure this list includes the one we want, or return an error.
+            for allocated_ip in &allocated_ips {
+                if allocated_ip.ip.ip() == external_ip
+                    && port_range
+                        .map(|(first, last)| {
+                            allocated_ip.first_port == SqlU16(first)
+                                && allocated_ip.last_port == SqlU16(last)
+                        })
+                        .unwrap_or(true)
```

Comment:
What does `port_range` being None mean here?

Reply:
There are two kinds of external IPs that have separate methods (`ensure_external_service_ip` and `ensure_external_service_snat_ip`) that both want to call this method. In the former case, all we have is an IP address, no port range, so when it calls this function it passes `None`. In the latter case, we have a `SourceNatConfig` value, which includes an IP and a port range, so when it calls this function it passes `Some(_)`.

In practice, the non-SNAT IPs are given a port range of `(0, 65535)`, but it seemed weird to hardcode that here when it's a decision being made elsewhere.

Reply:
Hmm interesting, should we use a type that's isomorphic to Option but not actually that? Definitely was surprised.

Reply:
(In particular, Ord doesn't make sense I think)

Reply:
Sorry, where is Ord coming into play?

Reply:
Sorry, what I mean is that if we define our own type, we can choose to not implement `Ord` on it, since it doesn't make sense in this case. We could also define other operations like `subset_of`/`superset_of`, which is what I think this code is trying to do.

Reply:
(But this is followup material, doesn't have to be done here.)

Reply:
Hmm, we don't want subset/superset here; this is just a straight equality check (except for having to convert to `SqlU16` to actually do the comparison). Is there a clearer way to express that?

Reply:
Ahh, no I think it was just that I misunderstood what this was doing :)